SmartDeviceLink/SDLIAPSession.m

```diff
@@ -66,17 +83,117 @@ - (BOOL)start {
 }
 
 - (void)stop {
-    [self stopStream:self.easession.outputStream];
-    [self stopStream:self.easession.inputStream];
+
+    if (!self.isDataSession) {
+        [self stopStream:self.easession.outputStream];
+        [self stopStream:self.easession.inputStream];
+    } else {
+        [self.ioStreamThread cancel];
+        
+        long lWait = dispatch_semaphore_wait(self.canceledSemaphore, dispatch_time(DISPATCH_TIME_NOW, streamThreadWaitSecs * NSEC_PER_SEC));
+        if (lWait == 0) {
+            [SDLDebugTool logInfo:@"Stream thread canceled"];
+        } else{
+           [SDLDebugTool logInfo:@"Error: failed to cancel stream thread"];
+        }
+        self.ioStreamThread = nil;
+        self.isDataSession = NO;
+    }
     self.easession = nil;
 }
 
+- (void)sendData:(NSData *)data {
+    // Enqueue the data for transmission on the IO thread
+    [self.sendDataQueue enqueueBuffer:data.mutableCopy];
```

Comment:
As in the original implementation, we want to consume the NSData bytes in chunks equal to the available space in the output stream. I don't know of a better way to do this unless the transport interface is changed to take a mutable NSData.